src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,230 @@
+        auto result = serializeV8(js, jsg::JsValue(value.getHandle(js)));
+        auto builder = callContext.initResults(capnp::MessageSize { result.size() / 8 + 8, 0 });
+        builder.initResult().setV8Serialized(kj::mv(result));
+      })));
+    });
+  }
+
+  KJ_DISALLOW_COPY_AND_MOVE(JsRpcTargetImpl);
+
+private:
+  // The following names are reserved by the Workers Runtime and cannot be called over RPC.
+  bool isReservedName(kj::StringPtr name) {
+    if (name == "fetch" ||
+        name == "connect" ||
+        name == "alarm" ||
+        name == "webSocketMessage" ||
+        name == "webSocketClose" ||
+        name == "webSocketError") {
+      return true;
+    }
+    return false;
+  }
+
+  // If the `methodName` is a known public method, we'll return it.
+  inline v8::Local<v8::Function> tryGetFn(
+      Worker::Lock& lock,
+      IoContext& ctx,
+      kj::StringPtr methodName) {
+
+    auto& handler = KJ_REQUIRE_NONNULL(lock.getExportedHandler(entrypointName, ctx.getActor()),
+        "Failed to get handler to worker.");
+
+    auto handle = handler.self.getHandle(lock);
+    auto methodStr = jsg::v8StrIntern(lock.getIsolate(), methodName);
+    auto fnHandle = jsg::check(handle->Get(lock.getContext(), methodStr));
+
+    jsg::Lock& js(lock);
+    v8::Local<v8::Object> obj = js.obj();
+    auto objProto = obj->GetPrototype().As<v8::Object>();
+
+    // Get() will check the Object and the prototype chain. We want to verify that the function
+    // we intend to call is not the one defined on the Object prototype.
+    bool isImplemented = fnHandle != jsg::check(objProto->Get(js.v8Context(), methodStr));
+
+    JSG_REQUIRE(isImplemented && fnHandle->IsFunction(), TypeError,
+        kj::str("The RPC receiver does not implement the method \"", methodName, "\"."));
+    JSG_REQUIRE(!isReservedName(methodName), TypeError,
+        kj::str("'", methodName, "' is a reserved method and cannot be called over RPC."));
+    return fnHandle.As<v8::Function>();
   }
+
+  // Deserializes the arguments and passes them to the given function.
+  v8::Local<v8::Value> invokeFn(
+      jsg::Lock& js,
+      v8::Local<v8::Function> fn,
+      kj::ArrayPtr<const kj::byte> serializedArgs) {
+    // We received arguments from the client, deserialize them back to JS.
+    if (serializedArgs.size() > 0) {
+      auto args = KJ_REQUIRE_NONNULL(
+          deserializeV8(js, serializedArgs).tryCast<jsg::JsArray>(),
+          "expected JsArray when deserializing arguments.");
+      // Call() expects a `Local<Value> []`... so we populate an array.
+      KJ_STACK_ARRAY(v8::Local<v8::Value>, arguments, args.size(), 8, 8);
+      for (size_t i = 0; i < args.size(); ++i) {
+        arguments[i] = args.get(js, i);
+      }
+      return jsg::check(fn->Call(js.v8Context(), fn, args.size(), arguments.begin()));
+    } else {
+      return jsg::check(fn->Call(js.v8Context(), fn, 0, nullptr));
```

Comment:
@MellowYarker @kentonv ... I think there's a bug here. The second argument to `fn->Call(...)` is supposed to the receiver (aka `this`)... which in this case should be the DO object and not `fn`. The way this is written `this` in these methods is going to equal the function and not the DO.

Reply:
PR: https://github.com/cloudflare/workerd/pull/1502